metrics/sinks/gcm/gcm.go

```diff
@@ -321,21 +325,43 @@ func CreateGCMSink(uri *url.URL) (core.DataSink, error) {
 		return nil, fmt.Errorf("invalid metrics parameter: %s", metrics)
 	}
 
-	if err := gce_util.EnsureOnGCE(); err != nil {
-		return nil, err
-	}
-
-	// Detect project ID
-	projectId, err := gce.ProjectID()
+	client, err := google.DefaultClient(oauth2.NoContext, gcm.MonitoringScope)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("error creating oauth2 client: %s", err)
 	}
 
-	// Create Google Cloud Monitoring service.
-	client := oauth2.NewClient(oauth2.NoContext, google.ComputeTokenSource(""))
 	gcmService, err := gcm.New(client)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("error creating GCM service: %s", err)
+	}
+
+	// Try and get the project ID from an environment variable first
+	var projectId string
+	projectId = os.Getenv(gcpProjectId)
+	if projectId == "" {
+		// If the variable is not set, move on to the default credentials file
+		file := os.Getenv(gcpCredentialEnv)
+		if file != "" {
+			// Attempt to load the config from the credentials file
+			conf, err := ioutil.ReadFile(file)
+			if err != nil {
+				return nil, fmt.Errorf("error reading default credentials file: %s", err)
+			}
+			var gcpConfig struct {
+				ProjectId string `json:"project_id"`
+			}
+			err = json.Unmarshal(conf, &gcpConfig)
+			if err != nil {
+				return nil, fmt.Errorf("error loading default credentials file: %s", err)
```

Comment:
That's fair -- the P12 format is listed as deprecated on the key download page.

I'll let the Kubernetes team comment on the refactoring, but keep in mind that extracting the project id from credentials is actually not GCE-specific (though it might be GCP-specific).